### ISSUE TITLE
Mock LDAP server state not shared among connections

### ISSUE THREAD

**Issue by brookskindle** [cannatag/ldap3#195]
Hello!

I've been trying to set up a mock LDAP server for a user authentication unit test and I've ran into a problem where I cannot create a user in one Connection and authenticate that user in another.
eg:
```python
server = Server(...)
conn1 = Connection(server, client_strategy=MOCK_SYNC)
# Create a user
conn1.strategy.add_entry("cn=testuser", {"userPassword":"mypassword"})

# Attempt to authenticate our newly created user
conn2 = Connection(server, client_strategy=MOCK_SYNC, user="cn=testuser", password="mypassword", authentication=SIMPLE)

conn2.bind()  # This fails
```
However, hacking `conn1` and setting the user, password, and authentication attributes on it and then calling `conn1.bind()` works just fine.

I did some digging and found that the Connection strategy object is unique per connection
```python
>>> conn1.strategy
<ldap3.strategy.mockSync.MockSyncStrategy object at 0xaaaaaaaaa>
>>> conn2.strategy
<ldap3.strategy.mockSync.MockSyncStrategy object at 0xffffffffffffff>
```
which doesn't make a whole lot of sense to me because the Server should be the one keeping track of the data, not the Connection.

This means that I can't create my user in a test function and then call my authenticate method unless I explicitly pass the connection object around.

IE this can't work:
```python
def test_authenticate_user():
  server = Server(...)
  conn = ...  # Same as above
  conn.strategy.add_entry(...)  # create user
  assert validate_user("testuser", "mypassword")
  assert not validate_user("invaliduser", "mypassword")

def validate_user(username, password):
  server = Server(...)
  conn = Connection(server, user=..., password=..., authentication=...)
  conn.bind()  # This will always return False because conn.strategy is a new object
```

Am I missing anything obvious in this assumption? Let me know!

Thanks,
Brooks

**[Edit]**: I'm using version 1.3.1

**Comment by brookskindle** [comment]
Giovanni,

Yes, that works for me, thank you for your help!

So long as I use the same `Server` object for my connections, the code works as I would expect it to.
ie:
```python
server = Server(...)
conn1 = Connection(server=server, ...)
conn1.add_entry(...)  # create a user
conn2 = Connection(server=server, user=..., password=...)
conn2.bind()  # returns True
```

For reference, I pip installed the dev branch at commit cc15c6b759b97add0f440230f37f01308e4233c1 (I haven't tried threading this in a threading context, so I can't comment on if that part works or not). Running pip freeze shows `ldap==1.3.3`

---

Do you think the state should be shared among all `Server` objects that share the same host? I know you can implicitly create a `Server` object when you create your `Connection` object by passing in a string for the server value.
```python
conn1 = Connection("ldap://my-server", ...)  # server implicitly created
conn2 = Connection("ldap://my-server", ...)  # new server with same host created
```
do you think it's

**Comment by cannatag** [comment]
Hi Brooks, the same server should have different instances on different ports so I think it's better to not match the server by its  name. Using the same Server object avoid this.

If you don't find any flaws in this implementation I will include it in the next release of ldap3.

Bye,
Giovanni

**Comment by brookskindle** [comment]
Giovanni,

That sounds fine with me, and I think this will actually make the clean up step easier when running tests - just delete and re-create the server object rather than figure out how to remove entries associated with the name.

Thanks again helping me out with this!

**Comment by cannatag** [comment]
released in 1.4.0
